File: apps/ios/Voiid/Voiid/Onboarding/OTPScreen.swift
//
//  OTPScreen.swift
//  Voiid
//
//  Enter verification code — the onboarding step after the phone number.
//
//  Built to the design source (`Voiid Ui/Screens/VerificationScreen.swift`) through
//  `OnboardingKit`, so the sizes, paddings and weights are the reference's own numbers.
//
//  ── THE BOXES ARE ONE FIELD WEARING SIX ─────────────────────────────────────────
//  Six separate `TextField`s is the obvious build and the wrong one. It breaks SMS autofill
//  (iOS hands the whole code to ONE field), needs hand-written focus-forwarding on every
//  keypress and backspace, and gets deleting-from-the-middle wrong in ways users find
//  immediately.
//
//  So there is ONE hidden field holding the whole string, and six boxes DRAWING it. Autofill,
//  paste, and backspace then work because they are the system's, not ours.
//
//  ── THIS SCREEN DOES VERIFY ─────────────────────────────────────────────────────
//  The design source checks nothing — its `onVerify` hands the digits up. The real work lives
//  here and is UNCHANGED by the restyle: Firebase verifies the code, the resulting ID token is
//  exchanged for OUR JWT, the verified number is persisted, E2E keys are published, and a
//  returning user with a server-side backup is offered restore before entering the app.
//
//  ── THE TIMER COUNTS SOMETHING REAL ─────────────────────────────────────────────
//  Unlike a launch progress bar, an expiry countdown is honest: the code really does expire.
//  `TimelineView` drives the tick rather than a `Timer` publisher, so a backgrounded app is not
//  burning a timer and the value shown on return is the TRUE remaining time rather than a stale
//  count that kept decrementing in the dark.
//
//  Committed to dark, like the rest of onboarding.
//

import SwiftUI

struct OTPScreen: View {
    @EnvironmentObject private var session: AppSession
    let onContinue: () -> Void
    /// Called instead of onContinue when the user already has a complete profile
    /// (returning user) — skip Signup/Profile and go straight to the app.
    var onExistingUser: () -> Void = {}
    /// Back to the phone step, to correct a wrong number.
    var onChangeNumber: (() -> Void)? = nil
    var phoneNumber: String = "+91 91234567890"
    /// E.164 number (no spaces) used for login. Defaults from `phoneNumber`.
    var e164: String? = nil
    /// Firebase verificationID from the Phone screen's sendCode.
    var verificationID: String = ""

    private let codeLength = 6
    /// How long a Firebase SMS code stays valid.
    private let validFor: TimeInterval = 120

    // Single source of truth: one hidden field holds all 6 digits; boxes display them.
    @State private var code = ""
    @State private var verifying = false
    @State private var resending = false
    @State private var errorText: String?
    /// Set when a returning user has a server-side backup — presents the restore
    /// flow before entering the app. nil = no backup (or not yet checked).
    @State private var restoreMeta: BackupMeta?
    /// The live verificationID. Starts as the one passed in and is REPLACED by a resend —
    /// verifying against the original id after a resend fails, because Firebase invalidates it.
    @State private var activeVerificationID = ""
    @State private var deadline = Date.now
    @FocusState private var focused: Bool

    private var isComplete: Bool { code.count == codeLength }

    /// Normalized E.164 ("+" + digits) for the login call.
    private var phoneE164: String {
        if let e164 { return e164 }
        let digits = phoneNumber.filter { $0.isNumber }
        return "+\(digits)"
    }

    var body: some View {
        ZStack {
            VoiidBrand.ground.ignoresSafeArea()

            ScrollView {
                VStack(spacing: 0) {
                    OnboardingHeader(
                        title: .stacked("Enter verification code", accent: "we sent to you"),
                        blurb: "We've sent a \(codeLength)-digit code to"
                    )

                    recipient
                        .padding(.top, 2)

                    codeBoxes
                        .padding(.top, VoiidSpacing.lg)

                    if let errorText {
                        Text(errorText)
                            .font(VoiidFont.rounded(13, .regular))
                            .foregroundColor(VoiidColor.error)
                            .multilineTextAlignment(.center)
                            .fixedSize(horizontal: false, vertical: true)
                            .padding(.top, VoiidSpacing.sm)
                    }

                    expiry
                        .padding(.top, VoiidSpacing.md)

                    securityNote
                        .padding(.top, VoiidSpacing.lg)

                    resendRow
                        .padding(.top, VoiidSpacing.lg)
                }
                .padding(.horizontal, VoiidSpacing.lg)
                .padding(.bottom, 170)
            }
            .scrollIndicators(.hidden)
            .onTapGesture { focused = false }

            VStack(spacing: 0) {
                Spacer(minLength: 0)
                OnboardingFooter {
                    OnboardingKitButton(title: "Verify & Continue",
                                        enabled: isComplete && !verifying) {
                        focused = false
                        Task { await verify() }
                    }
                }
            }
            .ignoresSafeArea(edges: .bottom)
        }
        .preferredColorScheme(.dark)
        .navigationBarTitleDisplayMode(.inline)
        .toolbarBackground(.hidden, for: .navigationBar)
        .task {
            activeVerificationID = verificationID
            startCountdown()
            // Deferred for the same reason as the phone field: during a push the field is not in
            // the responder chain when the view first appears, so focusing there silently fails.
            try? await Task.sleep(for: .milliseconds(350))
            focused = true
        }
        .fullScreenCover(item: $restoreMeta) { meta in
            RestoreMessagesView(meta: meta) {
                restoreMeta = nil
                onExistingUser()
            }
        }
    }

    // MARK: Recipient

    /// The number, with a way back to change it.
    ///
    /// "Change" is here rather than only in the back button because a wrong number is the most
    /// likely reason the code never arrives, and the user should not have to work out that
    /// "back" is how they fix it.
    private var recipient: some View {
        HStack(spacing: VoiidSpacing.sm) {
            Text(phoneNumber)
                .font(VoiidFont.rounded(17, .semibold))
                .foregroundColor(VoiidBrand.text)

            if let onChangeNumber {
                Button {
                    Haptics.tap()
                    onChangeNumber()
                } label: {
                    Text("Change")
                        .font(VoiidFont.rounded(15, .medium))
                        .foregroundColor(VoiidBrand.lime)
                }
                .buttonStyle(PressableButtonStyle())
            }
        }
        .frame(maxWidth: .infinity)
    }

    // MARK: The code

    /// Six boxes drawing one field. See the header for why it is built this way.
    private var codeBoxes: some View {
        ZStack {
            // The real field, invisible but focusable and full-width so autofill has a target.
            // `.opacity(0.01)` rather than `.hidden()`: a fully hidden field cannot receive
            // focus, and a zero-opacity one still participates in autofill.
            TextField("", text: $code)
                .keyboardType(.numberPad)
                .textContentType(.oneTimeCode)     // this is what makes SMS autofill work
                .focused($focused)
                .opacity(0.01)
                .onChange(of: code) { _, new in
                    let filtered = String(new.filter(\.isNumber).prefix(codeLength))
                    if filtered != new {
                        code = filtered
                        return          // fires again with the clean value; act on that pass
                    }
                    // A typed digit that did not complete the code still deserves feedback.
                    if !filtered.isEmpty && filtered.count < codeLength { Haptics.selection() }
                    if filtered.count == codeLength {
                        focused = false
                        Haptics.soft()
                    }
                }

            HStack(spacing: 10) {
                ForEach(0..<codeLength, id: \.self) { index in
                    box(at: index)
                }
            }
            // The boxes are decoration; taps belong to the field beneath them.
            .allowsHitTesting(false)
        }
        // One tap target across the whole row, so a user aiming at any box lands in the field.
        .contentShape(Rectangle())
        .onTapGesture { focused = true }
        .accessibilityElement()
        .accessibilityLabel("Verification code")
        .accessibilityValue(code.isEmpty ? "Empty" : code.map(String.init).joined(separator: " "))
    }

    private func box(at index: Int) -> some View {
        let digits = Array(code)
        let filled = index < digits.count
        // The caret sits on the next empty box — or on the last one when the code is full, so a
        // complete code still shows where a backspace would land.
        let isCursor = focused && (index == min(digits.count, codeLength - 1))

        return RoundedRectangle(cornerRadius: 14, style: .continuous)
            .fill(VoiidColor.fieldFill)
            .overlay(
                RoundedRectangle(cornerRadius: 14, style: .continuous)
                    .stroke(
                        isCursor ? VoiidBrand.lime : VoiidColor.fieldBorder,
                        lineWidth: isCursor ? 2 : 1
                    )
            )
            .frame(height: 62)
            .overlay {
                if filled {
                    Text(String(digits[index]))
                        .font(VoiidFont.rounded(26, .semibold))
                        .foregroundColor(VoiidBrand.text)
                        .monospacedDigit()
                } else {
                    // An underscore, not an empty box: it shows a character is expected here.
                    Rectangle()
                        .fill(VoiidBrand.textDim.opacity(0.5))
                        .frame(width: 18, height: 2)
                        .offset(y: 12)
                }
            }
            .animation(.easeOut(duration: 0.15), value: isCursor)
            .animation(.easeOut(duration: 0.15), value: filled)
    }

    // MARK: Expiry

    /// The countdown. Real — see the header.
    private var expiry: some View {
        TimelineView(.periodic(from: .now, by: 1)) { context in
            expiryLabel(remaining: max(0, deadline.timeIntervalSince(context.date)))
        }
    }

    @ViewBuilder
    private func expiryLabel(remaining: TimeInterval) -> some View {
        Group {
            if remaining <= 0 {
                Text("The code has expired. Request a new one.")
                    .font(VoiidFont.rounded(14))
                    .foregroundColor(VoiidColor.error)
            } else {
                (
                    Text("The code will expire in ")
                        .foregroundColor(VoiidBrand.textDim)
                    + Text(formatted(remaining))
                        .foregroundColor(VoiidBrand.lime)
                        .monospacedDigit()
                )
                .font(VoiidFont.rounded(14))
            }
        }
        .frame(maxWidth: .infinity)
        .accessibilityLabel(remaining <= 0
            ? "The code has expired"
            : "Code expires in \(Int(remaining)) seconds")
    }

    /// mm:ss. Built by hand rather than with a DateFormatter because this is a DURATION, and
    /// formatting it as a time-of-day would localize into nonsense in some regions.
    private func formatted(_ t: TimeInterval) -> String {
        let total = max(0, Int(t.rounded(.up)))
        return String(format: "%02d:%02d", total / 60, total % 60)
    }

    private func startCountdown() {
        deadline = .now.addingTimeInterval(validFor)
    }

    // MARK: Security note

    private var securityNote: some View {
        HStack(spacing: VoiidSpacing.md) {
            Circle()
                .fill(VoiidBrand.lime.opacity(0.10))
                .frame(width: 46, height: 46)
                .overlay {
                    Image(systemName: "lock")
                        .font(.system(size: 19, weight: .medium))
                        .foregroundColor(VoiidBrand.lime)
                }

            VStack(alignment: .leading, spacing: 3) {
                Text("Keep your code secure")
                    .font(VoiidFont.rounded(15, .semibold))
                    .foregroundColor(VoiidBrand.text)

                Text("Never share your code with anyone.\nVoiid will never ask for it.")
                    .font(VoiidFont.rounded(12.5))
                    .foregroundColor(VoiidBrand.textDim)
                    .fixedSize(horizontal: false, vertical: true)
            }

            Spacer(minLength: 0)
        }
        .padding(.horizontal, VoiidSpacing.md)
        .padding(.vertical, 14)
        .background(VoiidBrand.card)
        .clipShape(RoundedRectangle(cornerRadius: VoiidRadius.lg, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: VoiidRadius.lg, style: .continuous)
                .stroke(VoiidBrand.hairline, lineWidth: 1)
        )
        .accessibilityElement(children: .combine)
    }

    // MARK: Resend

    private var resendRow: some View {
        HStack(spacing: VoiidSpacing.sm) {
            Image(systemName: "arrow.clockwise")
                .font(.system(size: 17, weight: .medium))
                .foregroundColor(VoiidBrand.lime)

            Text("Didn't receive the code?")
                .font(VoiidFont.rounded(15))
                .foregroundColor(VoiidBrand.textDim)

            Button {
                Haptics.tap()
                Task { await resend() }
            } label: {
                Text(resending ? "Sending…" : "Resend code")
                    .font(VoiidFont.rounded(15, .semibold))
                    .foregroundColor(VoiidBrand.lime)
            }
            .buttonStyle(PressableButtonStyle())
            .disabled(resending || verifying)
        }
        .frame(maxWidth: .infinity)
    }

    // MARK: Actions

    /// Ask Firebase for a new code.
    ///
    /// THE NEW verificationID REPLACES THE OLD ONE. Firebase invalidates the previous id when it
    /// issues a fresh code, so verifying against the id this screen was constructed with would
    /// fail for every user who tapped Resend. The old build's Resend button did nothing at all,
    /// which at least could not get this wrong.
    private func resend() async {
        guard !resending, !verifying else { return }
        resending = true
        errorText = nil
        do {
            let newID = try await FirebasePhoneAuth.sendCode(to: phoneE164)
            activeVerificationID = newID
            code = ""
            startCountdown()
            focused = true
            Haptics.success()
        } catch {
            errorText = error.localizedDescription
            Haptics.error()
        }
        resending = false
    }

    /// Verify the code with Firebase, then exchange the Firebase ID token for our
    /// JWT. (Firebase sent the SMS on the previous screen.)
    private func verify() async {
        guard !verifying else { return }
        verifying = true; errorText = nil
        do {
            let idToken = try await FirebasePhoneAuth.verify(verificationID: activeVerificationID,
                                                             code: code)
            let profileComplete = try await AuthService.shared.loginWithFirebase(idToken: idToken)
            // Persist the VERIFIED phone number. The server never stores it, so this is the
            // ONLY place the real number is known — without this, Settings shows a placeholder.
            AppSession.saveVerifiedPhone(phoneE164)
            // AppSession.init() already ran at cold launch with an empty key, so update the
            // LIVE profile too — otherwise the number only appears after the next relaunch.
            session.profile.phoneNumber = phoneE164
            // Publish this device's E2E identity + prekeys (needed for encrypted chat).
            try? await E2EManager.shared.bootstrap()
            Haptics.success()
            if profileComplete {
                // Returning user. If the account has a server-side backup and this
                // device doesn't already hold the master secret (fresh install wiped
                // the E2E keychain), offer restore BEFORE entering the app. No backup
                // (or already restored) → proceed straight through.
                if !BackupManager.shared.hasLocalSecret,
                   let meta = try? await BackupService.shared.fetchBackupMeta() {
                    restoreMeta = meta
                } else {
                    onExistingUser()
                }
            } else {
                onContinue()
            }
        } catch {
            errorText = (error as? APIError)?.errorDescription ?? error.localizedDescription
            Haptics.error()
        }
        verifying = false
    }
}

File: apps/android/app/src/main/java/com/voiid/app/onboarding/OtpScreen.kt
package com.voiid.app.onboarding

import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.filled.Lock
import com.voiid.app.ui.components.VoiidMotion
import com.voiid.app.ui.theme.VoiidRadius
import com.voiid.app.ui.theme.VoiidSpacing
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.autofill.ContentType
import androidx.compose.ui.semantics.contentType
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import com.voiid.app.ui.components.LocalVoiidHaptics
import com.voiid.app.ui.theme.VoiidColor
import com.voiid.app.ui.theme.VoiidFont

/** Onboarding — 6-digit OTP verification (Figma Screen-3). Port of `OTPScreen.swift`. */
@Composable
fun OtpScreen(
    session: com.voiid.app.model.AppSession,
    phoneE164: String,
    verificationId: String,
    onBack: () -> Unit,
    onContinue: () -> Unit,
    /** Called instead of [onContinue] when a RETURNING user verifies — routes through the
     *  Verified screen before entering the app. Mirrors iOS `onExistingUser`. */
    onVerifiedExistingUser: () -> Unit = {},
) {
    val haptics = LocalVoiidHaptics.current
    val focus = LocalFocusManager.current
    val context = androidx.compose.ui.platform.LocalContext.current
    val scope = rememberCoroutineScope()
    val length = 6
    // Firebase SMS codes stay valid for 120s (iOS `validFor`).
    val validForSeconds = 120
    var code by remember { mutableStateOf("") }
    var focused by remember { mutableStateOf(false) }
    var verifying by remember { mutableStateOf(false) }
    var resending by remember { mutableStateOf(false) }
    var errorText by remember { mutableStateOf<String?>(null) }
    // The LIVE verificationID. Starts as the one passed in and is REPLACED by a resend —
    // verifying against the original id after a resend fails, because Firebase invalidates it.
    var activeVerificationId by remember { mutableStateOf(verificationId) }
    // Wall-clock deadline for the current code; 0 = not started yet.
    var deadlineAt by remember { mutableStateOf(0L) }
    var remainingSeconds by remember { mutableIntStateOf(validForSeconds) }
    // Non-null once a returning user logs in AND their account has a server backup:
    // we show the restore flow instead of dropping straight into the app.
    var restoreMeta by remember { mutableStateOf<com.voiid.app.net.BackupService.BackupMeta?>(null) }
    val fr = remember { androidx.compose.ui.focus.FocusRequester() }
    val complete = code.length == length
    val phoneNumber = phoneE164

    fun startCountdown() {
        deadlineAt = System.currentTimeMillis() + validForSeconds * 1000L
        remainingSeconds = validForSeconds
    }

    // One tick per second while a code is live; settles at 0 when it expires.
    androidx.compose.runtime.LaunchedEffect(deadlineAt) {
        if (deadlineAt == 0L) return@LaunchedEffect
        while (isActive) {
            val remaining = ((deadlineAt - System.currentTimeMillis()) / 1000L).toInt().coerceAtLeast(0)
            remainingSeconds = remaining
            if (remaining <= 0) break
            kotlinx.coroutines.delay(1000)
        }
    }

    fun focusCodeField() {
        runCatching { fr.requestFocus() }
    }

    fun findActivity(): android.app.Activity? {
        var ctx = context
        while (ctx is android.content.ContextWrapper) {
            if (ctx is android.app.Activity) return ctx
            ctx = ctx.baseContext
        }
        return null
    }

    // Ask Firebase for a new code. THE NEW verificationId REPLACES THE OLD ONE — Firebase
    // invalidates the previous id when it issues a fresh code, so verifying against the id
    // this screen was constructed with would fail for every user who tapped Resend.
    fun resend() {
        if (resending || verifying) return
        resending = true; errorText = null
        scope.launch {
            try {
                val activity = findActivity()
                    ?: throw IllegalStateException("Could not start verification from this screen.")
                val newId = com.voiid.app.net.FirebasePhoneAuth.sendCode(activity, phoneNumber)
                activeVerificationId = newId
                code = ""
                startCountdown()
                focusCodeField()
                haptics.success()
            } catch (e: Exception) {
                errorText = e.message ?: "Couldn't send a new code. Try again."
                haptics.error()
            }
            resending = false
        }
    }

    // Verify the code with Firebase, then exchange the Firebase ID token for our
    // JWT. (Firebase sent the SMS on the previous screen.)
    fun verify() {
        if (verifying) return
        verifying = true; errorText = null
        scope.launch {
            try {
                val idToken = com.voiid.app.net.FirebasePhoneAuth.verify(activeVerificationId, code)
                val profileComplete = session.auth.loginWithFirebase(idToken)
                // The verified number is the ONLY place we ever see it: the API never returns
                // a phone number (privacy). Keep it on our own row so Settings can show it.
                session.updateProfile(phoneE164 = phoneNumber)
                session.loadProfile()
                // Publish this device's E2E identity + prekeys (needed for chat).
                runCatching { com.voiid.app.net.E2EManager.get(context).bootstrap() }
                haptics.success()
                // Returning user (profile already complete) → offer to restore chats
                // if a server backup exists, else through Verified into the app.
                // New user → continue to Signup.
                if (profileComplete) {
                    restoreMeta = com.voiid.app.net.BackupService.BackupMeta()
                } else onContinue()
            } catch (e: Exception) {
                errorText = (e as? com.voiid.app.net.ApiError)?.userMessage ?: "Invalid or expired code."
                haptics.error()
            }
            verifying = false
        }
    }

    // A returning user with a server backup → show the restore flow, then continue
    // through Verified into the app.
    restoreMeta?.let { m ->
        RestoreFlow(
            session = session,
            meta = m,
            onDone = onVerifiedExistingUser,
            onSkip = onVerifiedExistingUser,
        )
        return
    }

    OnboardingScaffold(
        onBack = onBack,
        footer = {
            OnboardingKitButton(
                title = "Verify & Continue",
                enabled = complete && !verifying,
                busy = verifying,
            ) { focus.clearFocus(); verify() }
        },
    ) {
        Spacer(Modifier.height(VoiidSpacing.lg))

        OnboardingHeader(
            title = OnboardingTitleSpec.Stacked("Enter verification code", "we sent to you"),
            blurb = "We've sent a $length-digit code to",
        )

        // The recipient, with an explicit way back to correct it — a wrong number is the
        // most likely reason a code never arrives.
        Row(
            modifier = Modifier.fillMaxWidth().padding(top = 2.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp, Alignment.CenterHorizontally),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(phoneNumber, style = VoiidFont.rounded(17, FontWeight.SemiBold),
                color = VoiidBrand.text)
            Text(
                "Change",
                style = VoiidFont.rounded(15, FontWeight.Medium),
                color = VoiidBrand.lime,
                modifier = Modifier.clickable(
                    interactionSource = remember { MutableInteractionSource() },
                    indication = null,
                    enabled = !resending && !verifying,
                ) { haptics.tap(); onBack() },
            )
        }

        // Hidden field captures all input; decorationBox renders the display boxes.
        BasicTextField(
            value = code,
            onValueChange = { newVal ->
                val digits = newVal.filter(Char::isDigit).take(length)
                if (digits != code) {
                    code = digits
                    // A typed digit that did not complete the code still deserves feedback;
                    // completing it dismisses the keyboard with a soft press.
                    if (digits.isNotEmpty() && digits.length < length) haptics.selection()
                    if (digits.length == length) { focus.clearFocus(); haptics.soft() }
                }
            },
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
            // Text + cursor are invisible; the boxes below render the digits.
            textStyle = TextStyle(color = Color.Transparent),
            cursorBrush = SolidColor(Color.Transparent),
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = VoiidSpacing.lg)
                .focusRequester(fr)
                // ONE-TIME-CODE semantics: the Android counterpart of iOS's
                // `.textContentType(.oneTimeCode)`.
                .semantics { contentType = ContentType.Companion.SmsOtpCode }
                .onFocusChanged { focused = it.isFocused },
            decorationBox = { innerTextField ->
                Box {
                    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                        for (i in 0 until length) OtpBox(i, code, focused, length)
                    }
                    // Transparent input overlay keeps the IME attached + captures taps.
                    Box(Modifier.matchParentSize()) { innerTextField() }
                }
            },
        )
        LaunchedFocus(fr)

        // Start the expiry clock for the code that was sent on the previous screen.
        androidx.compose.runtime.LaunchedEffect(Unit) { startCountdown() }

        errorText?.let {
            Text(it, style = VoiidFont.rounded(13), color = VoiidColor.error,
                modifier = Modifier.fillMaxWidth().padding(top = VoiidSpacing.sm),
                textAlign = androidx.compose.ui.text.style.TextAlign.Center)
        }

        // Live countdown for the current code — real deadline, one tick per second.
        if (deadlineAt != 0L) {
            val expired = remainingSeconds <= 0
            Text(
                buildAnnotatedString {
                    if (expired) {
                        withStyle(SpanStyle(color = VoiidColor.error)) {
                            append("The code has expired. Request a new one.")
                        }
                    } else {
                        withStyle(SpanStyle(color = VoiidBrand.textDim)) {
                            append("The code will expire in ")
                        }
                        val total = remainingSeconds.coerceAtLeast(0)
                        withStyle(SpanStyle(color = VoiidBrand.lime, fontFeatureSettings = "tnum")) {
                            append("%02d:%02d".format(total / 60, total % 60))
                        }
                    }
                },
                style = VoiidFont.rounded(14),
                textAlign = androidx.compose.ui.text.style.TextAlign.Center,
                modifier = Modifier.fillMaxWidth().padding(top = VoiidSpacing.md),
            )
        }

        // The security note. Present on iOS and previously missing here entirely.
        Row(
            Modifier
                .fillMaxWidth()
                .padding(top = VoiidSpacing.lg)
                .clip(RoundedCornerShape(VoiidRadius.lg))
                .background(VoiidBrand.card)
                .border(1.dp, VoiidBrand.hairline, RoundedCornerShape(VoiidRadius.lg))
                .padding(VoiidSpacing.md),
            horizontalArrangement = Arrangement.spacedBy(VoiidSpacing.md),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(
                Modifier.size(46.dp).clip(CircleShape)
                    .background(VoiidBrand.lime.copy(alpha = 0.10f)),
                contentAlignment = Alignment.Center,
            ) {
                Icon(Icons.Default.Lock, null, tint = VoiidBrand.lime,
                    modifier = Modifier.size(19.dp))
            }
            Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Text("Keep your code secure", style = VoiidFont.rounded(15, FontWeight.SemiBold),
                    color = VoiidBrand.text)
                Text("Never share your code with anyone.\nVoiid will never ask for it.",
                    style = VoiidFont.rounded(12.5f), color = VoiidBrand.textDim)
            }
        }

        // Resend asks Firebase for a fresh code and REPLACES the verification id.
        Row(
            modifier = Modifier.fillMaxWidth().padding(top = VoiidSpacing.lg),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp, Alignment.CenterHorizontally),
        ) {
            Icon(Icons.Default.Refresh, null, tint = VoiidBrand.lime,
                modifier = Modifier.size(17.dp))
            Text("Didn't receive the code?", style = VoiidFont.rounded(15),
                color = VoiidBrand.textDim)
            Text(
                if (resending) "Sending…" else "Resend code",
                style = VoiidFont.rounded(15, FontWeight.SemiBold),
                color = VoiidBrand.lime.copy(alpha = if (resending || verifying) 0.5f else 1f),
                modifier = Modifier.clickable(
                    interactionSource = remember { MutableInteractionSource() },
                    indication = null,
                    enabled = !resending && !verifying,
                ) { haptics.tap(); resend() },
            )
        }

        Spacer(Modifier.height(VoiidSpacing.xl))
    }
}

@Composable
private fun RowScope.OtpBox(i: Int, code: String, keyboardUp: Boolean, length: Int) {
    // The caret STAYS on the last box when the code is complete. Android previously cleared
    // it (`code.length < length`), so a finished code highlighted nothing and the screen read
    // as though focus had been lost.
    val isCursor = keyboardUp && i == minOf(code.length, length - 1)
    val filled = i < code.length
    val borderColor by androidx.compose.animation.animateColorAsState(
        if (isCursor) VoiidBrand.lime else VoiidBrand.fieldEdge,
        animationSpec = androidx.compose.animation.core.tween(150, easing = VoiidMotion.easeOut),
        label = "otpBorder",
    )
    Box(
        modifier = Modifier
            .weight(1f)
            .height(62.dp)
            .clip(RoundedCornerShape(14.dp))
            .background(VoiidBrand.field)
            .border(
                width = if (isCursor) 1.5.dp else 1.dp,
                color = borderColor,
                shape = RoundedCornerShape(14.dp),
            ),
        contentAlignment = Alignment.Center,
    ) {
        if (filled) {
            Text(
                code[i].toString(),
                style = VoiidFont.rounded(26, FontWeight.SemiBold)
                    .copy(fontFeatureSettings = "tnum"),
                color = VoiidBrand.text,
            )
        } else {
            // An UNDERSCORE, not an empty box: it says "a character goes here" where a bare
            // outline says only "a box is here".
            Box(
                Modifier
                    .offset(y = 12.dp)
                    .size(width = 18.dp, height = 2.dp)
                    .background(VoiidBrand.textDim.copy(alpha = 0.5f))
            )
        }
    }
}

@Composable
private fun LaunchedFocus(fr: androidx.compose.ui.focus.FocusRequester) {
    androidx.compose.runtime.LaunchedEffect(Unit) {
        // 350ms, matching iOS. The field must be attached before focus is requested, and a
        // keyboard that appears before the screen has settled reads as a jump.
        kotlinx.coroutines.delay(350)
        runCatching { fr.requestFocus() }
    }
}
